### PR TITLE
Sound library for App Lab and Game Lab - sounds upload

### PR DESCRIPTION
**Pull request by caleybrock**
Not shown in PR:
- Create an S3 bucket for sounds
- Upload mp3 and matching json metadata files (about 300 sounds)

In PR:
- Create a dashboard endpoint to access the sound files from S3
- Adapt a script to create json metadata files for each mp3 file
- Adapt the manifest script to create a sound manifest json file based on what's in the S3 bucket
- Include json manifest file in repro

More info in sound library spec (requires login): https://docs.google.com/a/code.org/document/d/1Oj--H-xwrK3u4A0L5ML73n8XDutsH5n2i8caGRc3NL8/edit?usp=sharing

### CODE REVIEW

**Comment by caleybrock** [comment]
These sounds are much harder to search for with words, so I just added categories for aliases. We can always go back and add more if needed.

**Comment by caleybrock** [comment]
I'll make some inline notes too, but both scripts are basically the animation scripts, except replace `animations` with `sounds` and replace png with mp3.

File: tools/scripts/generateSoundMetadata.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    # Load each file and generate a metadata file for it
+    Parallel.map(unique_files) do |filename|
+      dirname = File.dirname(filename)
+      basename = File.basename(filename, '.mp3')
+      TagLib::MPEG::File.open(filename) do |mp3_file|
+        metadata = {}
+        metadata['name'] = basename
+        metadata['aliases'] = @options[:aliases] || []
+        metadata['time'] = mp3_file.audio_properties.length
```
Comment (caleybrock) [inline tools/scripts/generateSoundMetadata.rb:45]: TagLib allows us to get the duration of the mp3 files in seconds. Many of the Kenney.nl sounds are less than 1 second and show up as zero but I had trouble finding a ruby library that parsed sound in miliseconds, so I'll go with this for now.

TagLib library: http://robinst.github.io/taglib-ruby/

File: tools/scripts/generateSoundMetadata.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+require 'optparse'
+require 'parallel'
+require_relative '../../deployment'
+require_relative '../../lib/cdo/cdo_cli'
+include CdoCli
+
+begin
+  require 'taglib'
+rescue LoadError
+  puts "You need taglib installed locally to run this script. Run 'brew install taglib' then 'gem install taglib-ruby' and then add 'gem 'taglib-ruby'' to Gemfile and run bundle install. Do not check in changes to Gemfile."
```
Comment (islemaster) [inline tools/scripts/generateSoundMetadata.rb]: I think all that's necessary is `brew install taglib` and `gem install taglib-ruby`.  You shouldn't actually need to add anything to the gemfile to use it here.

File: tools/scripts/rebuildSoundLibraryManifest.rb [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    # Write result to file
+    File.open(DEFAULT_OUTPUT_FILE, 'w') do |file|
+      file.write(
+        JSON.pretty_generate(
+          {
+            # JSON-style file comment
+            '//': [
+              'sound Library Manifest',
+              'GENERATED FILE: DO NOT MODIFY DIRECTLY',
+              'See tools/scripts/rebuildsoundLibraryManifest.rb for more information.'
```
Comment (islemaster) [inline tools/scripts/rebuildSoundLibraryManifest.rb]: Nit: `s/rebuildsound/rebuildSound`

File: tools/scripts/rebuildSoundLibraryManifest.rb [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    info "Mapped #{alias_map.size} aliases."
+
+    # Write result to file
+    File.open(DEFAULT_OUTPUT_FILE, 'w') do |file|
+      file.write(
+        JSON.pretty_generate(
+          {
+            # JSON-style file comment
+            '//': [
+              'sound Library Manifest',
```
Comment (islemaster) [inline tools/scripts/rebuildSoundLibraryManifest.rb]: Nit: `s/sound/Sound`

File: tools/scripts/rebuildSoundLibraryManifest.rb [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        next <<-WARN
+There was an error parsing #{name}.json:
+#{json_error}
+The sound has been skipped.
+        WARN
+      end
+
+      # Verify that the parsed metadata contains all expected fields
+      next "sound #{name} is missing the 'name' attribute." unless metadata['name'].is_a?(String)
+      next "sound #{name} is missing the 'sound' attribute." unless metadata['time'].is_a?(Integer)
```
Comment (islemaster) [inline tools/scripts/rebuildSoundLibraryManifest.rb]: I think this warning should say it's missing the 'time' attribute, not the 'sound' attribute.

File: tools/scripts/rebuildSoundLibraryManifest.rb [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+      EOS
+
+      metadata
+    end
+    metadata_progress_bar.finish unless metadata_progress_bar.nil?
+    sound_metadata_by_name
+  end
+
+  # Given a metadata map, build the alias map
+  def build_alias_map(sound_metadata)
```
Comment (islemaster) [inline tools/scripts/rebuildSoundLibraryManifest.rb:323]: I think there's nothing sound-specific in this method.  Can we somehow extract it and share it between the sound manifest and animation manifest scripts?

File: tools/scripts/rebuildSoundLibraryManifest.rb [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+  def warn(s)
+    puts(s)
+  end
+end
+
+# Parse command-line options and then start the rebuild process
+options = {}
+cli_parser = OptionParser.new do |opts|
+  opts.banner = "Usage: ./rebuildsoundLibraryManifest.rb [options]"
```
Comment (islemaster) [inline tools/scripts/rebuildSoundLibraryManifest.rb]: Nit: `s/rebuildsound/rebuildSound`

**Pull request merged**